### PR TITLE
Improve `git restore` completions

### PR DESCRIPTION
**Pull request by hexium310**
The completions of `git restore` that are shown are changed by the options.

- `--source` or `-s`
  The result of `git ls-files` is shown.
- `--staged` or `-S`
  The staged files are shown.
- other
  The unstaged files are shown.

https://git-scm.com/docs/git-restore

### CODE REVIEW

**Comment by chitoku-k** [comment]
@hexium310 
Hi. Thanks for your contribution. Generally speaking, there should not be many functions that have similar features such as `_fzf_complete_git-staged-files` or `_fzf_complete_git-unstaged-files` where their difference is not quite noticeable at first glance. How about creating a function like `_fzf_complete_git-status-files` that takes what kind of files it prints as an argument?

**Comment by hexium310** [comment]
I think that it is good.

File: src/completers/git.zsh [hunk 1]
```diff
@@ -395,7 +425,7 @@ _fzf_complete_git-unstaged-files() {
                     -v red=${fg[red]} \
                     -v reset=$reset_color '
                         '$_fzf_complete_awk_functions'
-                        /^.[^ ]/ {
+                        /'${regexes[$files_kind]}'/ {
```
Comment (chitoku-k) [inline src/completers/git.zsh:428]: Do not interpolate a variable into the midst of an awk program. Pass the `$state` via `-v` and use that in conjunction with regex like the following:
```
state == "staged" && /^[^ ]/ {
    // ...
}
```

File: src/completers/git.zsh [hunk 2]
```diff
@@ -375,10 +402,13 @@ _fzf_complete_git-ls-files_post() {
...
 }
 
-_fzf_complete_git-unstaged-files() {
-    local git_options=$1
-    local fzf_options=$2
-    shift 2
+_fzf_complete_git-status-files() {
+    local -A regexes
+    regexes=(staged '^[^ ]' unstaged '^.[^ ]')
+    local files_kind=$1
```
Comment (chitoku-k) [inline src/completers/git.zsh:408]: Rename this to `state` as described in manual.

File: src/completers/git.zsh [hunk 3]
```diff
@@ -152,7 +169,17 @@ _fzf_complete_git() {
                 ;;
 
             *)
-                _fzf_complete_git-ls-files '' '--multi' $@
+                if [[ ${${(Q)${(z)arguments}}[(r)--source]} = '--source' ]] || [[ ${${(Q)${(z)arguments}}[(r)-s]} = '-s' ]]; then
```
Comment (chitoku-k) [inline src/completers/git.zsh:172]: Just check for whether the expansion results in a non-empty string. I'm not sure why there have been already two ways to do something like this written in this `git.zsh` though. Correct me if I'm wrong but this implementation looks like it doesn't accept `-Xs` where `X` is any other short option.
```suggestion
                if [[ -n ${${(Q)${(z)arguments}}[(r)--source]} ]] || [[ -n ${${(Q)${(z)arguments}}[(r)-s]} ]]; then
```

File: src/completers/git.zsh [hunk 4]
```diff
@@ -152,7 +169,17 @@ _fzf_complete_git() {
                 ;;
 
             *)
-                _fzf_complete_git-ls-files '' '--multi' $@
+                if [[ -n ${${(Q)${(z)arguments}}[(r)--source]} ]] || [[ -n ${${(Q)${(z)arguments}}[(r)-[^-]#s]} ]]; then
```
Comment (chitoku-k) [inline src/completers/git.zsh:172]: Question: Does it accept `-sX` as well?

Comment (hexium310) [reply src/completers/git.zsh:172]: No, don't accept. You can make sure by running below command.

```sh
echo ${$(echo -Xs)[(r)-[^-]#s]}
echo ${$(echo -sX)[(r)-[^-]#s]}
```

Comment (chitoku-k) [reply src/completers/git.zsh:172]: Then here needs to be fixed to accept it.

**Comment by chitoku-k** [comment]
From documentation it states that the `-s` / `--source` takes a tree as its required argument. The current implementation is broken.

**Comment by chitoku-k** [comment]
As it has been broken since that got merged, I found that the completion was incorrect in the following case whereas the long option is correct:
```zsh
$ git restore -s **
```

**Comment by chitoku-k** [comment]
OMG, it was I who made this buggy completion!

**Comment by chitoku-k** [comment]
Crucial!

**Pull request merged**